src/utils/contracts.ts

```diff
@@ -1,42 +1,82 @@
-import { ethers } from "ethers";
+import { BigNumber, ethers } from "ethers";
 import { ComposableCoW, ComposableCoW__factory } from "../types";
 import {
   COMPOSABLE_COW_CONTRACT_ADDRESS,
+  MAX_UINT32,
+  PollResultCode,
+  PollResultErrors,
   SupportedChainId,
 } from "@cowprotocol/cow-sdk";
+import { getLogger } from "./logging";
 
 // Selectors that are required to be part of the contract's bytecode in order to be considered compatible
 const REQUIRED_SELECTORS = [
   "cabinet(address,bytes32)",
   "getTradeableOrderWithSignature(address,(address,bytes32,bytes),bytes,bytes32[])",
 ];
 
-const CUSTOM_ERROR_ABI_MAP = {
-  PROOF_NOT_AUTHED: "ProofNotAuthed()",
-  SINGLE_ORDER_NOT_AUTHED: "SingleOrderNotAuthed()",
-  SWAP_GUARD_RESTRICTED: "SwapGuardRestricted()",
-  INVALID_HANDLER: "InvalidHandler()",
-  INVALID_FALLBACK_HANDLER: "InvalidFallbackHandler()",
-  INTERFACE_NOT_SUPPORTED: "InterfaceNotSupported()",
-  ORDER_NOT_VALID: "OrderNotValid(string)",
-  POLL_TRY_NEXT_BLOCK: "PollTryNextBlock(string)",
-  POLL_NEVER: "PollNever(string)",
-  POLL_TRY_AT_BLOCK: "PollTryAtBlock(uint256,string)",
-  POLL_TRY_AT_EPOCH: "PollTryAtEpoch(uint256,string)",
+// Define an enum for the custom error revert hints that can be returned by the ComposableCoW's interfaces.
+export enum CustomErrorSelectors {
+  PROOF_NOT_AUTHED = "PROOF_NOT_AUTHED",
+  SINGLE_ORDER_NOT_AUTHED = "SINGLE_ORDER_NOT_AUTHED",
+  SWAP_GUARD_RESTRICTED = "SWAP_GUARD_RESTRICTED",
+  INVALID_HANDLER = "INVALID_HANDLER",
+  INVALID_FALLBACK_HANDLER = "INVALID_FALLBACK_HANDLER",
+  INTERFACE_NOT_SUPPORTED = "INTERFACE_NOT_SUPPORTED",
+  ORDER_NOT_VALID = "ORDER_NOT_VALID",
+  POLL_TRY_NEXT_BLOCK = "POLL_TRY_NEXT_BLOCK",
+  POLL_NEVER = "POLL_NEVER",
+  POLL_TRY_AT_BLOCK = "POLL_TRY_AT_BLOCK",
+  POLL_TRY_AT_EPOCH = "POLL_TRY_AT_EPOCH",
+}
+
+type ParsedCustomError = {
+  [K in CustomErrorSelectors]: K extends
+    | CustomErrorSelectors.PROOF_NOT_AUTHED
+    | CustomErrorSelectors.SINGLE_ORDER_NOT_AUTHED
+    | CustomErrorSelectors.INTERFACE_NOT_SUPPORTED
+    | CustomErrorSelectors.INVALID_FALLBACK_HANDLER
+    | CustomErrorSelectors.INVALID_HANDLER
+    | CustomErrorSelectors.SWAP_GUARD_RESTRICTED
+    ? { selector: K }
+    : K extends
+        | CustomErrorSelectors.ORDER_NOT_VALID
+        | CustomErrorSelectors.POLL_TRY_NEXT_BLOCK
+        | CustomErrorSelectors.POLL_NEVER
+    ? { selector: K; message: string }
+    : K extends
+        | CustomErrorSelectors.POLL_TRY_AT_BLOCK
+        | CustomErrorSelectors.POLL_TRY_AT_EPOCH
+    ? { selector: K; message: string; blockNumberOrEpoch: number }
+    : never;
+}[CustomErrorSelectors];
```

Comment:
🤯

Reply:
Agree, this feels too complicated. What is this type needed for? 

Why not the function just returns a `PollResultErrors` so you also simplify the othe switch.

Reply:
This type covers the exact interface for all `ComposableCoW`-compatible orders (as returned from ABI on-chain). We control this interface, and therefore we can assert that _only_ these revert types are valid. If an order fails to adhere to this interface, it is to be dropped.

Typechain _does not_ do generation of typings for solidity custom error reverts (despite them being in the ABI), and therefore, we've basically had to implement this ourselves, which is essentially what this type is.